From 41c087a1ccdfd1827ceeea5464acf39c225a1eaf Mon Sep 17 00:00:00 2001
From: "Kirill A. Korinsky" <kirill@korins.ky>
Date: Mon, 12 Jun 2023 13:04:08 +0200
Subject: [PATCH] arm/Darwin: disable futex on BigSur

Unconditional enable of futex on macOS on arm had happened at
4ce2310cc5d9db835423168cac0d28ed8caff3eb, unfortunately it brokes build
on macOS 11 aka BigSur as:

  fatal error encountered in SBCL pid 42995 pthread 0x16b513000:
  GC invariant lost, file "thread.c", line 256

Building with --without-sb-futex won't help because it's enforced on all
arm darwin machines.

Here I've moved enabling it into make-config.sh on macOS 12 or never.
---
 make-config.sh       | 7 +++++--
 src/cold/shared.lisp | 2 +-
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git make-config.sh make-config.sh
index 0be42f451..0812b0ca9 100755
--- make-config.sh
+++ make-config.sh
@@ -628,9 +628,9 @@ case "$sbcl_os" in
         ;;
     darwin)
         printf ' :unix :bsd :darwin :mach-o' >> $ltf
+        darwin_version=`uname -r`
+        darwin_version_major=${DARWIN_VERSION_MAJOR:-${darwin_version%%.*}}
         if [ $sbcl_arch = "x86-64" ]; then
-            darwin_version=`uname -r`
-            darwin_version_major=${DARWIN_VERSION_MAJOR:-${darwin_version%%.*}}
 
             if (( 8 < $darwin_version_major )); then
 	        printf ' :inode64' >> $ltf
@@ -638,6 +638,9 @@ case "$sbcl_os" in
         fi
         if [ $sbcl_arch = "arm64" ]; then
             printf ' :darwin-jit :gcc-tls' >> $ltf
+            if (( 20 < $darwin_version_major )); then
+	        printf ' :sb-futex' >> $ltf
+            fi
         fi
         if $android; then
             echo "Android build is unsupported on darwin"
diff --git src/cold/shared.lisp src/cold/shared.lisp
index 695c168d6..241176aa0 100644
--- src/cold/shared.lisp
+++ src/cold/shared.lisp
@@ -291,7 +291,7 @@
              (arch (target-platform-keyword)))
         ;; Win32 conditionally adds :sb-futex in grovel-features.sh
         ;; Futexes aren't available in all macos versions, but they are available in all versions that support arm, so always enable them there
-        (when (target-featurep '(:and :sb-thread (:or :linux :freebsd :openbsd (:and :darwin :arm64))))
+        (when (target-featurep '(:and :sb-thread (:or :linux :freebsd :openbsd)))
           (pushnew :sb-futex sb-xc:*features*))
         ;; If may not be the best idea to put clock_gettime calls around every
         ;; futex_wait if the OS/libc you're building for does not have a vdso entry point.
-- 
2.41.0

